### PR TITLE
SONARJAVA-3191 S1905 should also raise issues in method invocation and constructor call arguments

### PR DESCRIPTION
**Pull request by amelie-renard-sonarsource**

### CODE REVIEW

File: java-checks/src/main/java/org/sonar/java/checks/RedundantTypeCastCheck.java [hunk 1]
```diff
@@ -57,14 +58,16 @@ public void visitNode(Tree tree) {
...
-    Type target = targetType(typeCastTree);
     Type expressionType = typeCastTree.expression().symbolType();
     if (isPrimitiveWrapperInConditional(expressionType, typeCastTree) || requiredForMemberAccess(typeCastTree)) {
       // Primitive wrappers excluded because covered by S2154
       return;
     }
-    if (target != null && (isRedundantNumericalCast(cast, expressionType) || isUnnecessarySubtypeCast(expressionType, typeCastTree, target))) {
+    Type cast = typeCastTree.type().symbolType();
+    Type target = targetType(typeCastTree);
+    boolean isArgument = skipParentheses(typeCastTree.parent()).is(Tree.Kind.ARGUMENTS);
```
Comment (m-g-sonar) [inline java-checks/src/main/java/org/sonar/java/checks/RedundantTypeCastCheck.java]: You only need `isArgument` within method `isUnnecessarySubtypeCast()`, and you can computed it from there, as you are anyway using passing `typeCastTree` as parameter.

File: java-checks/src/main/java/org/sonar/java/checks/RedundantTypeCastCheck.java [hunk 2]
```diff
@@ -114,16 +117,10 @@ private static Type targetType(TypeCastTree typeCastTree) {
...
-        // FIXME(SONARJAVA-3191) dead code
-        // If placed in invocation arguments, parent of typecast will be ARGUMENTS, not METHOD_INVOCATION
-        MethodInvocationTree mit = (MethodInvocationTree) parent;
-        if (mit.symbol().isMethodSymbol()) {
-          Symbol.MethodSymbol sym = (Symbol.MethodSymbol) mit.symbol();
-          int castArgIndex = mit.arguments().indexOf(typeCastTree);
-          return sym.parameterTypes().get(castArgIndex);
-        }
-        return null;
+      case ARGUMENTS:
```
Comment (m-g-sonar) [inline java-checks/src/main/java/org/sonar/java/checks/RedundantTypeCastCheck.java]: nice move, I didn't thought about this when creating the ticket! :+1:

File: java-checks/src/test/files/checks/RedundantTypeCastCheck.java [hunk 3]
```diff
@@ -18,16 +18,16 @@
...
     A a = (A) new B(); // Noncompliant {{Remove this unnecessary cast to "A".}}
     A[][] as = (A[][]) new B[1][1]; // Noncompliant {{Remove this unnecessary cast to "A[][]".}}
     B b;
     fun(b);
-    fun((A) b);
+    fun((A) b); // Compliant - exception to distinguish the method to call
     List<B> bees = new java.util.ArrayList<B>();
     java.util.List<A> aaas = (java.util.List) bees;
-    C c = new C((A)null);
+    C c = new C((A)null); // Compliant - exception to distinguish the constructor to call
```
Comment (m-g-sonar) [inline java-checks/src/test/files/checks/RedundantTypeCastCheck.java]: I would add a new constructor call to check that we are new raising issues in newClassTree as well.

File: java-checks/src/main/java/org/sonar/java/checks/RedundantTypeCastCheck.java [hunk 4]
```diff
@@ -114,16 +115,10 @@ private static Type targetType(TypeCastTree typeCastTree) {
...
+        Arguments arguments = (Arguments)parent;
+        int castArgIndex = arguments.indexOf(typeCastTree);
+        return arguments.get(castArgIndex).symbolType();
```
Comment (m-g-sonar) [inline java-checks/src/main/java/org/sonar/java/checks/RedundantTypeCastCheck.java]: Sorry I didn't catch it earlier... this won't work. What you wrote here is not equivalent to what we had earlier.
What is here now is this equivalent to do : `typeCastTree.symbolType()`, so the type of the resulting cast. We already have this information and that's not what we are searching for.

Previously, we were looking at the symbol of the method which was called, and looking at the EXPECTED type, as defined in his declaration.

In the following example: previously we would have retrieved `Collection`, today we are getting `ArrayList`

```
void foo(Collection param) { ... }
//       ^^^^^^^^^^ type from declaration = Collection

void bar(List param) {
  foo((ArrayList) param);
//     ^^^^^^^^^ type from cast = ArrayList
}
```

So in fact probably you will still need to get to the NewClassTree or MethodInvocationTree, but the symbol corresponding is only one parent ahead.

File: java-checks/src/main/java/org/sonar/java/checks/RedundantTypeCastCheck.java [hunk 5]
```diff
@@ -114,16 +115,10 @@ private static Type targetType(TypeCastTree typeCastTree) {
...
-        // If placed in invocation arguments, parent of typecast will be ARGUMENTS, not METHOD_INVOCATION
-        MethodInvocationTree mit = (MethodInvocationTree) parent;
-        if (mit.symbol().isMethodSymbol()) {
-          Symbol.MethodSymbol sym = (Symbol.MethodSymbol) mit.symbol();
-          int castArgIndex = mit.arguments().indexOf(typeCastTree);
-          return sym.parameterTypes().get(castArgIndex);
-        }
-        return null;
+      case ARGUMENTS:
+        Arguments arguments = (Arguments)parent;
```
Comment (m-g-sonar) [inline java-checks/src/main/java/org/sonar/java/checks/RedundantTypeCastCheck.java]: missing parenthesis after the cast.

File: java-checks/src/main/java/org/sonar/java/checks/RedundantTypeCastCheck.java [hunk 6]
```diff
@@ -57,14 +58,14 @@ public void visitNode(Tree tree) {
...
-    Type target = targetType(typeCastTree);
     Type expressionType = typeCastTree.expression().symbolType();
     if (isPrimitiveWrapperInConditional(expressionType, typeCastTree) || requiredForMemberAccess(typeCastTree)) {
       // Primitive wrappers excluded because covered by S2154
       return;
     }
-    if (target != null && (isRedundantNumericalCast(cast, expressionType) || isUnnecessarySubtypeCast(expressionType, typeCastTree, target))) {
+    Type cast = typeCastTree.type().symbolType();
+    Type target = targetType(typeCastTree);
+    if (target != null &&(isRedundantNumericalCast(cast, expressionType) || isUnnecessarySubtypeCast(expressionType, typeCastTree, target))) {
```
Comment (m-g-sonar) [inline java-checks/src/main/java/org/sonar/java/checks/RedundantTypeCastCheck.java]: missing parenthesis after `&&` . this line is unchanged.

File: java-checks/src/main/java/org/sonar/java/checks/RedundantTypeCastCheck.java (hunk 4 above)
Comment (amelie-renard-sonarsource) [reply java-checks/src/main/java/org/sonar/java/checks/RedundantTypeCastCheck.java]: You're totally right. Should be fixed

File: java-checks/src/main/java/org/sonar/java/checks/RedundantTypeCastCheck.java [hunk 7]
```diff
@@ -114,14 +116,14 @@ private static Type targetType(TypeCastTree typeCastTree) {
...
-          return sym.parameterTypes().get(castArgIndex);
+      case ARGUMENTS:
+        Arguments arguments = (Arguments) parent;
+        if (arguments.parent().is(Tree.Kind.METHOD_INVOCATION)) {
+          MethodInvocationTree mit = (MethodInvocationTree) arguments.parent();
+          return targetTypeFromMethodSymbol(mit.symbol(), mit.arguments(), typeCastTree);
+        } else if (arguments.parent().is(Tree.Kind.NEW_CLASS)) {
+          NewClassTree nct = (NewClassTree) arguments.parent();
+          return targetTypeFromMethodSymbol(nct.constructorSymbol(), nct.arguments(), typeCastTree);
         }
```
Comment (m-g-sonar) [inline java-checks/src/main/java/org/sonar/java/checks/RedundantTypeCastCheck.java]: It can be even more simplified in order to avoid calling 4 times `arguments.parent()`, and reusing `arguments`. It also makes code a bit easier to read:
```java
Arguments arguments = (Arguments) parent;
Tree invocation = arguments.parent();
if (invocation.is(Tree.Kind.METHOD_INVOCATION)) {
  return targetTypeFromMethodSymbol(((MethodInvocationTree) invocation).symbol(), arguments, typeCastTree);
}
// no need for else, early return
if (invocation.is(Tree.Kind.NEW_CLASS)) {
  return targetTypeFromMethodSymbol(((NewClassTree) invocation).constructorSymbol(), arguments, typeCastTree);
}
return null;
```
In term of code coverage, the only 3rd case of Arguments tree being used is when using annotations (see `AnnotationTree`). It worth adding a test case (if not already present). Note that for me it's OK if we don't consider these cases, as annotations parameters are always constant expression... I don't see many cases where wrong/redundant type cast would occur

```java
public class A {

  @MyAnnotation((int) (0L + 42))
             // ^^^^^^^^^^^^^^^ type cast in arguments from AnnotationTree
  Object field;

  @interface MyAnnotation {
    int value() default 42;
  }
}
```

File: java-checks/src/main/java/org/sonar/java/checks/RedundantTypeCastCheck.java [hunk 8]
```diff
@@ -53,17 +55,17 @@
 
   @Override
   public void visitNode(Tree tree) {
-    if(!hasSemantic()) {
+    if (!hasSemantic()) {
```
Comment (m-g-sonar) [inline java-checks/src/main/java/org/sonar/java/checks/RedundantTypeCastCheck.java]: unnecessary change :smile:

File: java-checks/src/main/java/org/sonar/java/checks/RedundantTypeCastCheck.java [hunk 9]
```diff
@@ -53,17 +55,17 @@
...
     }
     TypeCastTree typeCastTree = (TypeCastTree) tree;
-    Type cast = typeCastTree.type().symbolType();
-    Type target = targetType(typeCastTree);
     Type expressionType = typeCastTree.expression().symbolType();
     if (isPrimitiveWrapperInConditional(expressionType, typeCastTree) || requiredForMemberAccess(typeCastTree)) {
       // Primitive wrappers excluded because covered by S2154
       return;
     }
+    Type cast = typeCastTree.type().symbolType();
```
Comment (m-g-sonar) [inline java-checks/src/main/java/org/sonar/java/checks/RedundantTypeCastCheck.java]: :+1:

**Comment by m-g-sonar** [comment]
LGTM!
